### PR TITLE
zh: add usage for expression filter

### PR DESCRIPTION
**Pull request by lance6716**
### What is changed, added, or deleted? (Required)

 add doc for https://github.com/pingcap/dm/pull/1782

### Which DM version(s) do your changes apply to? (Required)

- [x] master (the latest development version)
- [x] v2.0 (TiDB DM 2.0 versions)
- [ ] v1.0 (TiDB DM 1.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [x] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: zh/feature-expression-filter.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+对于 `expr_filter` 库下的 `tbl` 表，当插入偶数的 `c` (`c % 2 = 0`) 时，不将这条插入同步到下游。
+
+## 配置规则
+
+- schema: 要匹配的上游数据库库名，不支持通配符匹配或正则匹配
+- table: 要匹配的上游表名，不支持通配符匹配或正则匹配
+- insert-value-expr: 可以配置一个表达式，对 INSERT 类型的 binlog event (WRITE_ROWS_EVENT) 带有的值生效。不能与 update-old-value-expr、update-new-value-expr、delete-value-expr 出现在一个配置项中。
+- update-old-value-expr: 可以配置一个表达式，对 UPDATE 类型的 binlog event (UPDATE_ROWS_EVENT) 更新对应的旧值生效。不能与 insert-value-expr、delete-value-expr 出现在一个配置项中。
+- update-new-value-expr: 可以配置一个表达式，对 UPDATE 类型的 binlog event (UPDATE_ROWS_EVENT) 更新对应的旧值生效。不能与 insert-value-expr、delete-value-expr 出现在一个配置项中。
```
Comment (kennytm) [inline zh/feature-expression-filter.md:48]:
```suggestion
- update-new-value-expr: 可以配置一个表达式，对 UPDATE 类型的 binlog event (UPDATE_ROWS_EVENT) 更新对应的新值生效。不能与 insert-value-expr、delete-value-expr 出现在一个配置项中。
```

File: zh/feature-expression-filter.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    expression-filters: ["even_c"]
+
+expression-filter:
+  even_c:
+    schema: "expr_filter"
+    table: "tbl"
+    insert-value-expr: "c % 2 = 0"
+```
+
+上面的配置中配置了 `even_c` 规则，并让 source ID 为 `mysql-replica-01` 的上游引用了改规则。`even_c` 规则的含义是：
```
Comment (kennytm) [inline zh/feature-expression-filter.md:38]:
```suggestion
上面的配置中配置了 `even_c` 规则，并让 source ID 为 `mysql-replica-01` 的上游引用了该规则。`even_c` 规则的含义是：
```

**Comment by CharLotteiu** [comment]
The changes in this PR are related to DM v2.0.5. Does it need to be merged after v2.0.5 releases? @lance6716

**Comment by lance6716** [comment]
> The changes in this PR are related to DM v2.0.5. Does it need to be merged after v2.0.5 releases? @lance6716

a user has used the nightly (patched) version, so I think the dev branch could be updated at this time.

File: zh/feature-expression-filter.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+> - 当二者同时配置时，会将更新旧值满足 update-old-value-expr **且**更新新值满足 update-new-value-expr 的行变动过滤掉
+> - 当只配置一者时，配置的这条表达式会决定是否过滤**整个行变更**，即对旧值的删除和新值的插入会作为一个整体被过滤掉
+
+SQL 表达式可以涉及一列、多列、使用 TiDB 支持的 SQL 函数。例如 `c % 2 = 0`，`a*a + b*b = c*c`、`ts > NOW()`
+
+TIMESTAMP 类型的默认时区是 UTC。可以使用 `c_timestamp = '2021-01-01 12:34:56.5678+08:00'` 的方式显式指定时区。
+
+配置项 `expression-filter` 下可以配置多种过滤规则，上游数据源在 `expression-filters` 中引用这些规则使其生效。当有多条规则生效时，匹配到**任意一条**规则即会导致某个行变更被过滤。
+
+> **注意：**
```
Comment (fubinzh) [inline zh/feature-expression-filter.md:85]: SQL 表达式过滤 跟 全量导入、导出无关，这个可能需要澄清一下。

Comment (lance6716) [reply zh/feature-expression-filter.md:89]: https://github.com/pingcap/docs-dm/pull/729/commits/9a885fc6b6be6eb672845da2753ac20303f6ac86

added here

**Comment by CharLotteiu** [comment]
> > The changes in this PR are related to DM v2.0.5. Does it need to be merged after v2.0.5 releases? @lance6716
> 
> an user has used the nightly (patched) version, so I think the dev branch could be updated at this time.

Well then...how about removing the `v2.0.5` in this PR and only mentioning the version in the release 2.0 branch? And after this PR is cherry-picked to release 2.0, can we wait for DM v2.0.5 before merging that PR?

File: zh/feature-expression-filter.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+expression-filter:
+  even_c:
+    schema: "expr_filter"
+    table: "tbl"
+    insert-value-expr: "c % 2 = 0"
+```
+
+上面的配置中配置了 `even_c` 规则，并让 source ID 为 `mysql-replica-01` 的上游引用了该规则。`even_c` 规则的含义是：
+
+对于 `expr_filter` 库下的 `tbl` 表，当插入偶数的 `c` (`c % 2 = 0`) 时，不将这条插入同步到下游。
```
Comment (Liuxiaozhen12) [inline zh/feature-expression-filter.md:44]:
```suggestion
对于 `expr_filter` 库下的 `tbl` 表，当插入的 `c` 为偶数 (`c % 2 = 0`) 时，不将这条插入同步到下游。
```

File: zh/feature-expression-filter.md [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+```
+
+## 配置规则
+
+- schema: 要匹配的上游数据库库名，不支持通配符匹配或正则匹配
+- table: 要匹配的上游表名，不支持通配符匹配或正则匹配
+- insert-value-expr: 可以配置一个表达式，对 INSERT 类型的 binlog event (WRITE_ROWS_EVENT) 带有的值生效。不能与 update-old-value-expr、update-new-value-expr、delete-value-expr 出现在一个配置项中
+- update-old-value-expr: 可以配置一个表达式，对 UPDATE 类型的 binlog event (UPDATE_ROWS_EVENT) 更新对应的旧值生效。不能与 insert-value-expr、delete-value-expr 出现在一个配置项中
+- update-new-value-expr: 可以配置一个表达式，对 UPDATE 类型的 binlog event (UPDATE_ROWS_EVENT) 更新对应的新值生效。不能与 insert-value-expr、delete-value-expr 出现在一个配置项中
+- delete-value-expr: 可以配置一个表达式，对 DELETE 类型的 binlog event (DELETE_ROWS_EVENT) 带有的值生效。不能与 insert-value-expr、update-old-value-expr、update-new-value-expr 出现在一个配置项中
```
Comment (Liuxiaozhen12) [inline zh/feature-expression-filter.md:74]:
```suggestion
- insert-value-expr: 配置一个表达式，对 INSERT 类型的 binlog event (WRITE_ROWS_EVENT) 带有的值生效。不能与 update-old-value-expr、update-new-value-expr、delete-value-expr 出现在一个配置项中
- update-old-value-expr: 配置一个表达式，对 UPDATE 类型的 binlog event (UPDATE_ROWS_EVENT) 更新对应的旧值生效。不能与 insert-value-expr、delete-value-expr 出现在一个配置项中
- update-new-value-expr: 配置一个表达式，对 UPDATE 类型的 binlog event (UPDATE_ROWS_EVENT) 更新对应的新值生效。不能与 insert-value-expr、delete-value-expr 出现在一个配置项中
- delete-value-expr: 配置一个表达式，对 DELETE 类型的 binlog event (DELETE_ROWS_EVENT) 带有的值生效。不能与 insert-value-expr、update-old-value-expr、update-new-value-expr 出现在一个配置项中
```

File: zh/feature-expression-filter.md [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+> **注意：**
+>
+> update-old-value-expr 可以与 update-new-value-expr 同时配置。
+>
+> - 当二者同时配置时，会将更新旧值满足 update-old-value-expr **且**更新新值满足 update-new-value-expr 的行变动过滤掉
+> - 当只配置一者时，配置的这条表达式会决定是否过滤**整个行变更**，即对旧值的删除和新值的插入会作为一个整体被过滤掉
```
Comment (Liuxiaozhen12) [inline zh/feature-expression-filter.md:81]:
```suggestion
> - 当只配置一者时，配置的这条表达式会决定是否过滤**整个行变更**，即旧值的删除和新值的插入会作为一个整体被过滤掉
```

File: zh/feature-expression-filter.md [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+SQL 表达式可以涉及一列、多列、使用 TiDB 支持的 SQL 函数。例如 `c % 2 = 0`，`a*a + b*b = c*c`、`ts > NOW()`
```
Comment (Liuxiaozhen12) [inline zh/feature-expression-filter.md:83]:
```suggestion
SQL 表达式可以涉及一列、多列，也可使用 TiDB 支持的 SQL 函数。例如 `c % 2 = 0`，`a*a + b*b = c*c`、`ts > NOW()`
```

File: zh/feature-expression-filter.md [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+为了解决上面的问题，DM 可以使用 SQL 表达式过滤某些行变更。DM 所要求的 ROW 模式的 binlog 会在 binlog event 中带有所有列的值。基于这些值，用户可以配置 SQL 表达式，如果该表达式对于某条行变更的计算结果是 `TRUE`，DM 就不会向下游同步该条行变更。
+
+> **注意：**
+>
+> 该功能只会在增量复制阶段生效，并不会在全量迁移阶段生效。
+
+## 配置示例
+
+表达式过滤在 task 配置文件里面与 [Binlog Event Filter](key-features.md#binlog-event-filter) 类似，详见下面配置样例。完整的配置及意义，可以参考 [DM 完整配置文件示例](task-configuration-file-full.md#完整配置文件示例)：
```
Comment (Liuxiaozhen12) [inline zh/feature-expression-filter.md:19]:
```suggestion
表达式过滤在数据迁移任务配置文件里面与 [Binlog Event Filter](key-features.md#binlog-event-filter) 类似，详见下面配置样例。完整的配置及意义，可以参考 [DM 完整配置文件示例](task-configuration-file-full.md#完整配置文件示例)：
```

File: zh/feature-expression-filter.md [hunk 9]
```diff
@@ -0,0 +1,91 @@
+---
+title: 使用 SQL 表达式过滤某些行变更
+---
+
+# 使用 SQL 表达式过滤某些行变更
+
+## 概述
+
+在进行数据同步的过程中，DM 提供了 [Binlog Event Filter](key-features.md#binlog-event-filter) 过滤某些类型的 binlog event，例如不向下游同步 DELETE 以达到归档、审计等目的。但是 Binlog Event Filter 无法以更细粒度判断某一行的 DELETE 是否要被过滤。
```
Comment (CharLotteiu) [inline zh/feature-expression-filter.md:9]:
```suggestion
在数据迁移的过程中，DM 提供了 [Binlog Event Filter](key-features.md#binlog-event-filter) 功能过滤某些类型的 binlog event，例如不向下游迁移 `DELETE` 事件以达到归档、审计等目的。但是 Binlog Event Filter 无法以更细粒度判断某一行的 `DELETE` 事件是否要被过滤。
```

File: zh/feature-expression-filter.md [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+为了解决上面的问题，DM 可以使用 SQL 表达式过滤某些行变更。DM 所要求的 ROW 模式的 binlog 会在 binlog event 中带有所有列的值。基于这些值，用户可以配置 SQL 表达式，如果该表达式对于某条行变更的计算结果是 `TRUE`，DM 就不会向下游同步该条行变更。
```
Comment (CharLotteiu) [inline zh/feature-expression-filter.md:11]:
```suggestion
为了解决上述问题，DM 支持使用 SQL 表达式过滤某些行变更。DM 所要求的 ROW 模式的 binlog 会在 binlog event 中带有所有列的值。用户可以基于这些值配置 SQL 表达式。如果该表达式对于某条行变更的计算结果是 `TRUE`，DM 就不会向下游迁移该条行变更。
```

File: zh/feature-expression-filter.md [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    expression-filters: ["even_c"]
+
+expression-filter:
+  even_c:
+    schema: "expr_filter"
+    table: "tbl"
+    insert-value-expr: "c % 2 = 0"
+```
+
+上面的配置中配置了 `even_c` 规则，并让 source ID 为 `mysql-replica-01` 的上游引用了该规则。`even_c` 规则的含义是：
```
Comment (CharLotteiu) [inline zh/feature-expression-filter.md:42]:
```suggestion
上面的示例配置了 `even_c` 规则，并让 source ID 为 `mysql-replica-01` 的数据源引用了该规则。`even_c` 规则的含义是：
```

File: zh/feature-expression-filter.md [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+
+对于 `expr_filter` 库下的 `tbl` 表，当插入的 `c` 为偶数 (`c % 2 = 0`) 时，不将这条插入同步到下游。
```
Comment (CharLotteiu) [inline zh/feature-expression-filter.md:44]:
```suggestion
对于 `expr_filter` 库下的 `tbl` 表，当插入的 `c` 的值为偶数 (`c % 2 = 0`) 时，不将这条插入语句迁移到下游。
```

File: zh/feature-expression-filter.md [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+
+该示例的同步效果是：
+
+上游
```
Comment (CharLotteiu) [inline zh/feature-expression-filter.md:48]:
```suggestion
在上游数据源增量插入以下数据：
```

File: zh/feature-expression-filter.md [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+
+```sql
+INSERT INTO tbl(id, c) VALUES (1, 1), (2, 2), (3, 3), (4, 4);
+```
+
+下游
```
Comment (CharLotteiu) [inline zh/feature-expression-filter.md:54]:
```suggestion
随后在下游查询 `tbl` 表，可见只有 `c` 的值为单数的行迁移到了下游： 
```

File: zh/feature-expression-filter.md [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+- delete-value-expr: 配置一个表达式，对 DELETE 类型的 binlog event (DELETE_ROWS_EVENT) 带有的值生效。不能与 insert-value-expr、update-old-value-expr、update-new-value-expr 出现在一个配置项中
+
+> **注意：**
+>
+> update-old-value-expr 可以与 update-new-value-expr 同时配置。
+>
+> - 当二者同时配置时，会将更新旧值满足 update-old-value-expr **且**更新新值满足 update-new-value-expr 的行变动过滤掉
+> - 当只配置一者时，配置的这条表达式会决定是否过滤**整个行变更**，即旧值的删除和新值的插入会作为一个整体被过滤掉
+
+SQL 表达式可以涉及一列、多列，也可使用 TiDB 支持的 SQL 函数。例如 `c % 2 = 0`，`a*a + b*b = c*c`、`ts > NOW()`
```
Comment (CharLotteiu) [inline zh/feature-expression-filter.md:83]:
```suggestion
SQL 表达式可以涉及一列或多列，也可使用 TiDB 支持的 SQL 函数，例如 `c % 2 = 0`、`a*a + b*b = c*c`、`ts > NOW()`。
```

File: zh/feature-expression-filter.md [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+
+TIMESTAMP 类型的默认时区是 UTC。可以使用 `c_timestamp = '2021-01-01 12:34:56.5678+08:00'` 的方式显式指定时区。
+
+配置项 `expression-filter` 下可以配置多种过滤规则，上游数据源在 `expression-filters` 中引用这些规则使其生效。当有多条规则生效时，匹配到**任意一条**规则即会导致某个行变更被过滤。
```
Comment (CharLotteiu) [inline zh/feature-expression-filter.md:87]:
```suggestion
配置项 `expression-filter` 下可以定义多条过滤规则，上游数据源在其 `expression-filters` 配置项中引用需要的规则使其生效。当有多条规则生效时，匹配到**任意一条**规则即会导致某个行变更被过滤。
```

File: zh/feature-expression-filter.md [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+
+> **注意：**
+>
+> 为某张表设置过多的表达式过滤会增加 DM 的计算开销，可能导致同步变慢。
```
Comment (CharLotteiu) [inline zh/feature-expression-filter.md:91]:
```suggestion
> 为某张表设置过多的表达式过滤会增加 DM 的计算开销，可能导致数据迁移速度变慢。
```

File: zh/TOC.md [hunk 18]
```diff
@@ -17,6 +17,7 @@
       - [悲观模式](feature-shard-merge-pessimistic.md)
       - [乐观模式](feature-shard-merge-optimistic.md)
     - [迁移使用 GH-ost/PT-osc 的源数据库](feature-online-ddl-scheme.md)
+    - [使用 SQL 表达式过滤某些行变更](feature-expression-filter.md)
```
Comment (CharLotteiu) [inline zh/TOC.md:20]: 需不需要在[ Data Migration 简介](https://docs.pingcap.com/zh/tidb-data-migration/stable/overview#高级功能)中的高级功能一节也加一下相关描述？

File: zh/feature-expression-filter.md [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
++------+------+
+| id   | c    |
++------+------+
+|    1 |    1 |
+|    3 |    3 |
++------+------+
+2 rows in set (0.001 sec)
+```
+
+## 配置规则
```
Comment (CharLotteiu) [inline zh/feature-expression-filter.md:67]:
```suggestion
## 配置参数及规则说明
```

File: zh/task-configuration-file-full.md [hunk 20]
```diff
@@ -120,6 +126,7 @@ mysql-instances:
     route-rules: ["route-rule-1", "route-rule-2"]  # 该上游数据库实例匹配的表到下游数据库的 table routing 规则名称
     filter-rules: ["filter-rule-1", "filter-rule-2"]                # 该上游数据库实例匹配的表的 binlog event filter 规则名称
     block-allow-list:  "bw-rule-1"                 # 该上游数据库实例匹配的表的 block-allow-list 过滤规则名称，如果 DM 版本早于 v2.0.0-beta.2 则使用 black-white-list
+    expression-filters: ["even_c"]
```
Comment (CharLotteiu) [inline zh/task-configuration-file-full.md:129]: 这一行也请加一下注释说明～

File: zh/task-configuration-file-full.md [hunk 21]
```diff
@@ -75,6 +75,12 @@ filters:                                        # 上游数据库实例匹配的
     events: ["all dml"]
     action: Do
 
+expression-filter:
+  even_c:                            # 过滤 `expr_filter`.`tbl` 的 c 为偶数的插入
+    schema: "expr_filter"
+    table: "tbl"
+    insert-value-expr: "c % 2 = 0"
```
Comment (CharLotteiu) [inline zh/task-configuration-file-full.md:82]:
```suggestion
expression-filter:#定义数据源迁移行变更的过滤规则，可以定义多个规则
    # 过滤 `expr_filter`.`tbl` 的 c 为偶数的插入
    even_c:         #规则名称                   
    schema: "expr_filter" #要匹配的上游数据库库名，不支持通配符匹配或正则匹配
    table: "tbl" #要匹配的上游表名，不支持通配符匹配或正则匹配
    insert-value-expr: "c % 2 = 0"
```
在 GitHub 网页上改的，可能格式不太对。。

File: zh/feature-expression-filter.md [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+
+为了解决上面的问题，DM 可以使用 SQL 表达式过滤某些行变更。DM 所要求的 ROW 模式的 binlog 会在 binlog event 中带有所有列的值。基于这些值，用户可以配置 SQL 表达式，如果该表达式对于某条行变更的计算结果是 `TRUE`，DM 就不会向下游同步该条行变更。
+
+> **注意：**
+>
+> 该功能只会在增量复制阶段生效，并不会在全量迁移阶段生效。
+
+## 配置示例
+
+表达式过滤在数据迁移任务配置文件里面与 [Binlog Event Filter](key-features.md#binlog-event-filter) 类似，详见下面配置样例。完整的配置及意义，可以参考 [DM 完整配置文件示例](task-configuration-file-full.md#完整配置文件示例)：
```
Comment (CharLotteiu) [inline zh/feature-expression-filter.md:19]:
```suggestion
与 [Binlog Event Filter](key-features.md#binlog-event-filter) 类似，表达式过滤需要在数据迁移任务配置文件里配置，详见下面配置样例。完整的配置及意义，可以参考 [DM 完整配置文件示例](task-configuration-file-full.md#完整配置文件示例)：
```

File: zh/feature-expression-filter.md [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
++------+------+
+|    1 |    1 |
+|    3 |    3 |
++------+------+
+2 rows in set (0.001 sec)
+```
+
+## 配置规则
+
+- schema: 要匹配的上游数据库库名，不支持通配符匹配或正则匹配
```
Comment (CharLotteiu) [inline zh/feature-expression-filter.md:69]:
```suggestion
- `schema`：要匹配的上游数据库库名，不支持通配符匹配或正则匹配。
```

File: zh/feature-expression-filter.md [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+- table: 要匹配的上游表名，不支持通配符匹配或正则匹配
+- insert-value-expr: 配置一个表达式，对 INSERT 类型的 binlog event (WRITE_ROWS_EVENT) 带有的值生效。不能与 update-old-value-expr、update-new-value-expr、delete-value-expr 出现在一个配置项中
+- update-old-value-expr: 配置一个表达式，对 UPDATE 类型的 binlog event (UPDATE_ROWS_EVENT) 更新对应的旧值生效。不能与 insert-value-expr、delete-value-expr 出现在一个配置项中
+- update-new-value-expr: 配置一个表达式，对 UPDATE 类型的 binlog event (UPDATE_ROWS_EVENT) 更新对应的新值生效。不能与 insert-value-expr、delete-value-expr 出现在一个配置项中
+- delete-value-expr: 配置一个表达式，对 DELETE 类型的 binlog event (DELETE_ROWS_EVENT) 带有的值生效。不能与 insert-value-expr、update-old-value-expr、update-new-value-expr 出现在一个配置项中
```
Comment (CharLotteiu) [inline zh/feature-expression-filter.md:74]: ditto as L69

File: zh/feature-expression-filter.md [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
+
+> **注意：**
+>
+> update-old-value-expr 可以与 update-new-value-expr 同时配置。
+>
+> - 当二者同时配置时，会将更新旧值满足 update-old-value-expr **且**更新新值满足 update-new-value-expr 的行变动过滤掉
+> - 当只配置一者时，配置的这条表达式会决定是否过滤**整个行变更**，即旧值的删除和新值的插入会作为一个整体被过滤掉
```
Comment (CharLotteiu) [inline zh/feature-expression-filter.md:81]:
```suggestion
> - 当二者同时配置时，会将更新旧值满足 update-old-value-expr **且**更新新值满足 update-new-value-expr 的行变动过滤掉。
> - 当只配置一者时，配置的这条表达式会决定是否过滤**整个行变更**，即旧值的删除和新值的插入会作为一个整体被过滤掉。
```

File: zh/feature-expression-filter.md (hunk 23 above)
Comment (lance6716) [reply zh/feature-expression-filter.md:69]: 其他的我也改一下

**Review by Liuxiaozhen12 — APPROVED** [review]

**Review by CharLotteiu — APPROVED** [review]
rest LGTM

**Review by kennytm — APPROVED** [review]

**Review by lichunzhu — APPROVED** [review]
LGTM

**Pull request merged**